### ISSUE TITLE
Week 1-2: Immediate Fixes and Critical Tests

### ISSUE THREAD

**Issue by DavidZalewski** [DavidZalewski/chess#4]
1. Fix MultiDimensionalCacheTests.AddManyItems_PartitionsCorrectly()
2. Address the failing tests in TestAnalyzer.MethodsNeedingTestShouldBeLessThan3()
3. Start working on the most critical methods needing tests (e.g., AddActionInSequence, PlayActionSequence, Log)

**Comment by DavidZalewski** [comment]
1 is complete
2 has been worked on for several days - there are only 5 methods remaining from 12+
3. AddActionInSequence done, PlayActionSequence done, Log done, including other areas like Pawn Promotion Tracking, Saving and Loading of Game State, and other areas.

Overall: 90% complete

**Comment by DavidZalewski** [comment]
I have this class, it has a method annotated with the attribute [TestNeeded]
```
using Chess.Attributes;
using Chess.Board;
using Chess.GameState;
using Chess.Globals;
using Chess.Pieces;
using NUnit.Framework;

namespace Chess.Callbacks
{
    [Serializable]
    internal class SpecialMovesHandlers
    {
        public static Action<string>? GetActionFromResult;
        public static Func<string>? PawnPromotionPromptUser;
        public static bool ByPassPawnPromotionPromptUser { get; set; } = false;
        public static PromotionTracker promotionTracker = new PromotionTracker();

        [TestNeeded]
        public static bool DoCastleMove(ChessBoard cb, BoardPosition bp, ChessPiece king)
        {
            StaticLogger.Trace();
            string action = "";
            int hv = bp.FileAsInt;
            int vv = bp.RankAsInt;

            ChessPiece? rook = cb.GetActivePieces().Find(p => p.GetCurrentPosition() == bp);

            Assert.That(rook, Is.Not.Null);

            if (rook != null)
            {
                Assert.That(rook.GetPiece(), Is.EqualTo(ChessPiece.Piece.ROOK));

                // is king left of rook, or right of rook?
                int d = king.GetCurrentPosition().FileAsInt - rook.GetCurrentPosition().FileAsInt;
                FILE kh = king.GetCurrentPosition().File;
                FILE rh = rook.GetCurrentPosition().File;
                RANK v = king.GetCurrentPosition().Rank;
                // k=4, r=0 4-0=4, k=4, r2=8, k-r2=4-7 = -3
                if (d == 4)
                {
                    // queen side castle
                    // k goes -2 squares
                    // r goes +3 squares
                    action = "Castle (Queen Side)";
                    kh -= 2;
                    rh += 3;
                }
                else if (d == -3)
                {
                    // king side castle
                    // k goes +2 squares
                    // r goes -2 squares
                    action = "Castle (King Side)";
                    kh += 2;
                    rh -= 2;
                }
                else
                {
                    cb.GenerateBoardID();
                    throw new Exception("Unexpected Horizontal Distance Found when castling. Are you sure this is a valid castle? BoardID: " + cb.BoardID);
                }

                BoardPosition kingLastPosition = king.GetCurrentPosition();
                BoardPosition rookLastPosition = rook.GetCurrentPosition();

                // set board manually
                cb.SetBoardValue(kingLastPosition, 0);
                cb.SetBoardValue(rookLastPosition, 0);

                cb.SetPieceAtPosition(new(v, kh), king);
                cb.SetPieceAtPosition(new(v, rh), rook);
            }

            GetActionFromResult?.Invoke(action);

            return true;
        }
        
        [TestNeeded]
        public static bool IsEnPassantMove(ChessBoard chessBoard, BoardPosition boardPosition, ChessPiece pawnAttemptingEnPassant)
        {
            StaticLogger.Trace();
            RANK enPassantRow;
            ChessPiece.Color opponentColor;
            int enPassantOffSet = 0;

            if (pawnAttemptingEnPassant.GetColor().Equals(ChessPiece.Color.WHITE))
            {
                enPassantRow = RANK.FIVE;
                opponentColor = ChessPiece.Color.BLACK;
                enPassantOffSet = -1;
            }
            else
            {
                enPassantRow = RANK.FOUR;
                opponentColor = ChessPiece.Color.WHITE;
                enPassantOffSet = +1;
            }

            BoardPosition pawnPos = pawnAttemptingEnPassant.GetCurrentPosition();

            // Is the pawn in the correct Row to do this? En Passant is only possible if a pawn is on a specific row on the board
            if (pawnPos.Rank != enPassantRow)
                return false;

            // Are there opponent pieces to its immediate left or right?
            BoardPosition bpl = pawnPos.Left();
            BoardPosition bpr = pawnPos.Right();
            List<BoardPosition> boardPositionsToCheck = new();

            if (bpl != null)
                boardPositionsToCheck.Add(bpl);
            if (bpr != null)
                boardPositionsToCheck.Add(bpr);

            foreach (BoardPosition bpToCheck in boardPositionsToCheck)
            {
                // Is there an opponent piece at this position?
                if (chessBoard.IsPieceAtPosition(bpToCheck, opponentColor))
                {
                    ChessPiece? opponentPiece = chessBoard.GetActivePieces().Find((ChessPiece cp) => cp.GetCurrentPosition() == bpToCheck);
                    Assert.That(opponentPiece, Is.Not.Null, "This assertion failed. If ChessBoard.IsPieceAtPosition returns true, the piece must exist in the collection");

                    // Is the opponent piece a pawn?
                    if (opponentPiece is ChessPiecePawn)
                    {
                        // Did that pawn move 2 squares?
                        if ((opponentPiece as ChessPiecePawn).MovedTwoSquares)
                        {
                            // is the position a capture position?
                            // Get the opponent pawn position, and get the position that is 1 square behind it
                            BoardPosition oppPos = opponentPiece.GetCurrentPosition();
                            // this operation needs to support both + 1 (for black) and -1 (for white)
                            BoardPosition enPassantCapturePos = new((RANK)oppPos.RankAsInt + enPassantOffSet, oppPos.File);
                            if (enPassantCapturePos == boardPosition)
                            {
                                (opponentPiece as ChessPiecePawn).IsEnPassantTarget = true;
                                return true; // This is a valid En Passant capture move
                            }
                        }
                    }
                }

            }

            return false;
        }

        [TestNeeded]
        internal static void PawnPromotion(ChessBoard board, BoardPosition position, ChessPiece piece)
        {
            StaticLogger.Trace();
            string choice = "Q";
            if (PawnPromotionPromptUser != null && !ByPassPawnPromotionPromptUser)
                choice = PawnPromotionPromptUser.Invoke();

            Func<string, ChessPiece> switchReturnPiece = (string chosenPiece) =>
            {
                int newPieceID;
                // This fixes the issue with KingCheckService incrementing this counter with potential future moves
                if (ByPassPawnPromotionPromptUser)
                {
                    Console.WriteLine($"PawnPromotion: ByPassPawnPromotionPromptUser set - use 0 for newPieceID");
                    newPieceID = 0;
                }
                else
                {
                    newPieceID = promotionTracker.GetNextID(piece.GetColor(), chosenPiece.ToUpper());
                }
                Console.WriteLine($"PawnPromotion: newPieceID = {newPieceID} , choice: {choice}");
                return chosenPiece switch
                {
                    "q" => new ChessPieceQueen(piece.GetColor(), newPieceID, position),
                    "r" => new ChessPieceRook(piece.GetColor(), newPieceID, position),
                    "k" => new ChessPieceKnight(piece.GetColor(), newPieceID, position),
                    "b" => new ChessPieceBishop(piece.GetColor(), newPieceID, position),
                    _ => new ChessPieceQueen(piece.GetColor(), newPieceID, position),
                };
            };

            ChessPiece newPiece = switchReturnPiece(choice);
            board.AddPiece(newPiece);
        }
    }
}
```

These methods currently do not have unit tests written directly for them. But they are under test in an integration test that plays out many chess replays and verifies them against the engine. So this code is being targeted under test. Do you still feel its necessary to write unit tests for these, given that this code is already under coverage?

For example, see this test:
```
        [Test]
        public void PawnsOnly_KingInCheck_NoPawnPromotion()
        {
            // Arrange
            Queue<string> consoleInputs = new();
            consoleInputs.Enqueue("n"); // no to tutorial
            consoleInputs.Enqueue("PawnsOnly"); // PawnsOnly game
            consoleInputs.Enqueue("n"); // no to ai
            consoleInputs.Enqueue("WP3 C4");
            consoleInputs.Enqueue("BP7 G5");
            consoleInputs.Enqueue("WP3 C5");
            consoleInputs.Enqueue("BP4 D5");
            consoleInputs.Enqueue("WP3 D6"); // en passant
            consoleInputs.Enqueue("BP7 G4");
            consoleInputs.Enqueue("WP3 D7");
            consoleInputs.Enqueue("BK F8");
            consoleInputs.Enqueue("WP6 F4");
            consoleInputs.Enqueue("BP7 F3"); // en passant
            consoleInputs.Enqueue("WP1 A3");
            consoleInputs.Enqueue("BP7 F2");
            consoleInputs.Enqueue("WK D1");
            consoleInputs.Enqueue("quit");

            IConsole consoleService = new MockConsoleService(consoleInputs);
            GameController gameController = GetGameController(consoleService);
            GameManager game = new GameManager(consoleService, gameController);

            // override what GameManager constructor has registered for callback, with this one instead
            SpecialMovesHandlers.PawnPromotionPromptUser = () =>
            {
                consoleService.WriteLine("Pawn Promoted. Choose type: (Q - Queen, R - Rook, K - Knight, B - Bishop)");
                return "Q";
            };

            // Act
            game.Start();

            // Assert
            Assert.That(((MockConsoleService)consoleService).OutputContainsString("Pawn Promoted."), Is.False);
        }
```

Should we just remove the [TestNeeded] attribute from this class and call it a day, or should we implement tests for this class?

**Comment by DavidZalewski** [comment]
Your question touches on the nuances of testing strategies, particularly when integration tests already cover significant functionality. Here's a structured approach to decide whether to proceed with unit tests or to rely solely on existing integration tests:

### Consideration for Unit Testing:

1. **Isolation of Logic:**
   - **Pros for Unit Tests:**
     - Unit tests can isolate each method, testing them in complete isolation from the rest of the system. This allows for pinpointing exactly where a failure occurs, which can be invaluable for maintenance and refactoring.
     - They ensure that individual components work correctly even when the context changes, like updates to other parts of the system.

   - **Cons:**
     - Writing unit tests requires additional time and effort, especially for methods with complex logic or dependencies.

2. **Coverage and Specificity:**
   - **Integration Tests:**
     - Your integration test (`PawnsOnly_KingInCheck_NoPawnPromotion`) is comprehensive but tests many aspects at once. If there's an issue, it might be less clear which specific part failed.

   - **Unit Tests:**
     - Would provide detailed insight into how each method performs under various conditions, not just the scenarios played out in integration tests.

3. **Future-proofing:**
   - If these methods are likely to change or expand, unit tests can act as a safety net, ensuring that new changes do not break existing functionality.

4. **Complexity of the Methods:**
   - **DoCastleMove**: This method involves moving pieces, checking positions, and handling exceptions. The logic here is ripe for unit testing due to its complexity and the potential for subtle bugs.
   - **IsEnPassantMove**: Similarly, this method has specific conditions that could benefit from thorough unit testing to ensure all edge cases are covered.
   - **PawnPromotion**: Although tested indirectly, direct tests could verify promotion logic under different scenarios, especially the `switch` statement.

### Decision Points:

- **Existing Coverage:** Since you have integration tests that cover these methods within the context of an entire game, you might feel that additional unit tests are redundant if:
  - The integration tests are reliable, well-maintained, and cover all necessary scenarios for these methods.
  - The codebase is stable and not expected to undergo significant changes in these areas.

- **Value Added by Unit Tests:**
  - If there's a perceived risk in missing edge cases or if the methods are critical to the game's logic, writing unit tests could add value by:
    - Increasing confidence in each method's correctness.
    - Facilitating easier debugging and maintenance.

### Recommendation:

- **Remove [TestNeeded] Attribute?** 
  - Yes, if you decide that the current integration tests are sufficient for ensuring the functionality of these methods. However, keep in mind that:

- **Option to Write Unit Tests:**
  - If you have the resources, consider writing unit tests for:
    - **DoCastleMove**: Due to its complexity with position manipulation and potential for logical errors.
    - **IsEnPassantMove**: For its specific conditions which might not be fully covered by integration tests alone.

  This approach would provide a more granular view of your application's correctness, especially useful in complex logic scenarios.

- **Documentation**: Even if you choose not to write unit tests, consider improving documentation or adding comments to explain why these methods are considered tested enough through integration tests. This helps future developers understand the testing strategy.

In summary, if these methods are complex or pivotal to the game's mechanics, investing in unit tests might still be beneficial. However, if the integration tests are comprehensive and the methods are fairly straightforward or stable, you might opt to enhance documentation and rely on the integration tests alone.

**Comment by DavidZalewski** [comment]
Chess Application Starting
Would you like to play the tutorial? (y/n):
n
Select RuleSet:
*Classic* -- The game we all know and love
*PawnsOnly* -- No knights, bishops, rooks, or queens. Just a king and his pawns
*NuclearHorse* -- The game where the most powerful piece is also the most destructive, can you survive the environment?
*SevenByEight* -- The board is changed to be a 7 x 8 grid instead
*KingsForce* -- A pawn connected to a king can capture in all directions; the king can disable a square
Classic
*Adding RuleSet: Classic
Applying Rule Sets
Use AI for Black? (y, n)?
n
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|BK1|BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|BP5|BP6|BP7|BP8|8
6|   |   |   |   |   |   |   |   |8
5|   |   |   |   |   |   |   |   |8
4|   |   |   |   |   |   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|WP4|WP5|WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 1 - White to move. Please enter a command (piece name + position : ie. 'WK1 C3')
WP4 D4
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|BK1|BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|BP5|BP6|BP7|BP8|8
6|   |   |   |   |   |   |   |   |8
5|   |   |   |   |   |   |   |   |8
4|   |   |   |WP4|   |   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|   |WP5|WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 2 - Black to move. Please enter a command (piece name + position : ie. 'BP4 D5')
BP5 E5
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|BK1|BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|   |BP6|BP7|BP8|8
6|   |   |   |   |   |   |   |   |8
5|   |   |   |   |BP5|   |   |   |8
4|   |   |   |WP4|   |   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|   |WP5|WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 3 - White to move. Please enter a command (piece name + position : ie. 'WK1 C3')
WP4 BP5
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|BK1|BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|   |BP6|BP7|BP8|8
6|   |   |   |   |   |   |   |   |8
5|   |   |   |   |WP4|   |   |   |8
4|   |   |   |   |   |   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|   |WP5|WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 4 - Black to move. Please enter a command (piece name + position : ie. 'BP4 D5')
BK1 C6
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|   |BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|   |BP6|BP7|BP8|8
6|   |   |BK1|   |   |   |   |   |8
5|   |   |   |   |WP4|   |   |   |8
4|   |   |   |   |   |   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|   |WP5|WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 5 - White to move. Please enter a command (piece name + position : ie. 'WK1 C3')
WP5 E4
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|   |BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|   |BP6|BP7|BP8|8
6|   |   |BK1|   |   |   |   |   |8
5|   |   |   |   |WP4|   |   |   |8
4|   |   |   |   |WP5|   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|   |   |WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 6 - Black to move. Please enter a command (piece name + position : ie. 'BP4 D5')
BK1 WP4
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|   |BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|   |BP6|BP7|BP8|8
6|   |   |   |   |   |   |   |   |8
5|   |   |   |   |BK1|   |   |   |8
4|   |   |   |   |WP5|   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|   |   |WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 7 - White to move. Please enter a command (piece name + position : ie. 'WK1 C3')
WP5 BK1
Invalid Move. Please Try Again.
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*
8|BR1|   |BB1|BQ1|BK |BB2|BK2|BR2|8
7|BP1|BP2|BP3|BP4|   |BP6|BP7|BP8|8
6|   |   |   |   |   |   |   |   |8
5|   |   |   |   |BK1|   |   |   |8
4|   |   |   |   |WP5|   |   |   |8
3|   |   |   |   |   |   |   |   |8
2|WP1|WP2|WP3|   |   |WP6|WP7|WP8|8
1|WR1|WK1|WB1|WQ1|WK |WB2|WK2|WR2|8
*|*A*|*B*|*C*|*D*|*E*|*F*|*G*|*H*|*

Turn 7 - White to move. Please enter a command (piece name + position : ie. 'WK1 C3')
quit
Invalid command. Please enter a command.
GameManager: Ending Game

**Comment by DavidZalewski** [comment]
The engine is saying the quit command is invalid, but then it ends the game. Something is being triggered when it shouldn't be.
